scipy/spatial/distance.py

```diff
@@ -329,6 +329,13 @@ def _validate_vector(u, dtype=None):
     return u
 
 
+def _validate_weights(w, dtype=None):
+    w = _validate_vector(w, dtype=dtype)
+    if np.any(w < 0):
+        raise ValueError("Input weights should be all non-negative")
```

Comment:
Shouldn't this be `not np.all(w >= 0)`, to behave well on `nan`?

Reply:
Depends, should nans propagate or not? Right now I see metrics are inconsistent in that respect, e.g. ``euclidean`` gives a ValueError`` if there's a nan while ``cosine`` just returns ``nan``.

Reply:
Thinking again, either seems fine, but consistency would obviously be good. The error message is technically correct here, since nan is non-negative!